### PR TITLE
Create DeepCopy() and Json Marshal/Unmarshal for proto-public

### PR DESCRIPTION
**Pull request by thisisnotashwin**
### Description

In order to make Kubernetes CRDs easy to build, maintain, and test, we would like to embed the proto types we are building CRDs for directly as the spec of the CRD. A few changes were necessary in order for that to happen and are detailed below.

• Overwrite JSON Marshal/Unmarshal: The `json` tags on the generated proto.pb.go files are snake_cased. The equivalent CRD fields needs to be camelCased though since this is in line with Kubernetes best practices. The `protobuf` tags for said resources though, use camelCase tags for their JSON marshaling. By generating files that overwrite the default JSON Marshal/Unmarshal to use the protobuf JSON marshaler/unmarshaler, we can get around the deserialization problem and use camelCase in our CRD definitions for the keys.
To ensure we still support snake_case keys, we were able to create resources that used snake_case keys directly against the http api and it worked.

• Generate DeepCopyObject() for resources: For kubernetes to correctly serialize and deserialize a CRD, all fields on said CRD need to implement a `DeepCopyInto()` method. A generator was created in order to build the same for our proto-public resources.

• Kubebuilder tags: Enum types in proto translate into `int32` types in golang. This leads to the CRDs generated for the fields to only accept int values for those enum fields. Using explicit kubebuilder tags allow us to get around that by explicitly allowing string types for those fields as well as declaring them as enums so that non-supported values get rejected by the API as a pre-processing step. Additional kubebuilder tags were necessary in order to serialize the pbstruct type which is effectively a schemaless type.

### Testing & Reproduction steps

CRDs created using these protos here: https://github.com/hashicorp/consul-k8s/pull/3019

### PR Checklist

* [x] not a security concern

### CODE REVIEW

File: proto-public/pbmesh/v2beta1/tcp_route.proto [hunk 1]
```diff
@@ -10,9 +10,7 @@ import "pbresource/annotations.proto";
 
 // NOTE: this should align to the GAMMA/gateway-api version, or at least be
 // easily translatable.
-//
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/tcp_route.proto:13]: Can you restore the extra blank comment lines that were removed?

File: proto-public/pbmesh/v2beta1/routing_json.gen.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Code generated by protoc-json-shim. DO NOT EDIT.
+package meshv2beta1
+
+import (
+	protojson "google.golang.org/protobuf/encoding/protojson"
+)
+
+var (
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/routing_json.gen.go:8]: Why did it produce this basically empty file?

File: proto-public/pbmesh/v2beta1/routing_deepcopy.gen.go [hunk 3]
```diff
@@ -0,0 +1,2 @@
+// Code generated by protoc-gen-deepcopy. DO NOT EDIT.
+package meshv2beta1
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/routing_deepcopy.gen.go:2]: Why did it produce this basically empty file?

File: proto-public/pbmesh/v2beta1/http_route.proto [hunk 4]
```diff
@@ -167,17 +173,14 @@ message HTTPQueryParamMatch {
   // Name is the name of the HTTP query param to be matched. This must be an
   // exact string match. (See
   // https://tools.ietf.org/html/rfc7230#section-2.7.3).
-  //
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/http_route.proto:170]: Could you restore a few more of these empty comment lines?

File: proto-public/pbmesh/v2beta1/grpc_route.proto [hunk 5]
```diff
@@ -41,23 +41,18 @@ message GRPCRouteRule {
   repeated GRPCRouteFilter filters = 2;
 
   // BackendRefs defines the backend(s) where matching requests should be sent.
-  //
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/grpc_route.proto:44]: Could you restore a few more of these empty comment lines?

File: proto-public/pbmesh/v2beta1/computed_routes.proto [hunk 6]
```diff
@@ -151,7 +150,6 @@ enum BackendTargetDetailsType {
   BACKEND_TARGET_DETAILS_TYPE_DIRECT = 1;
 
   // Indirect means that the target is not directly routable from a route.
-  //
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/computed_routes.proto:154]: Could you restore a few more of these empty comment lines?

File: proto-public/pbmesh/v2beta1/computed_routes.proto [hunk 7]
```diff
@@ -62,7 +62,6 @@ message ComputedHTTPRouteRule {
 message ComputedHTTPBackendRef {
   // BackendTarget indicates which key in the targets map provides
   // the rest of the configuration.
-  //
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/computed_routes.proto:65]: Could you restore a few more of these empty comment lines?

File: proto-public/pbmesh/v2beta1/common.proto [hunk 8]
```diff
@@ -9,24 +9,22 @@ import "pbresource/resource.proto";
...
 }
 
 message BackendReference {
-  // For east/west configuration, this should point to a pbcatalog.Service.
+  // For east/west configuration, this should point to a Service.
   hashicorp.consul.resource.Reference ref = 1;
 
   // For east/west this is the name of the Consul Service port to direct traffic to
   // or empty to imply using the same value as the parent ref.
-  //
```
Comment (rboyer) [inline proto-public/pbmesh/v2beta1/common.proto:29]: Could you restore a few more of these empty comment lines?

File: proto-public/pbauth/v2beta1/traffic_permissions.proto [hunk 9]
```diff
@@ -7,27 +7,21 @@ import "pbresource/annotations.proto";
 message TrafficPermissions {
   option (hashicorp.consul.resource.spec) = {scope: SCOPE_NAMESPACE};
 
-  // destination is a configuration of the destination proxies
+  // Destination is a configuration of the destination proxies
   // where these traffic permissions should apply.
   Destination destination = 1;
 
   // Action can be either allow or deny for the entire object. It will default to allow.
-  //
```
Comment (rboyer) [inline proto-public/pbauth/v2beta1/traffic_permissions.proto:15]: Could you restore a few more of these empty comment lines?

File: proto-public/pbcatalog/v2beta1/protocol_deepcopy.gen.go [hunk 10]
```diff
@@ -0,0 +1,2 @@
+// Code generated by protoc-gen-deepcopy. DO NOT EDIT.
+package catalogv2beta1
```
Comment (rboyer) [inline proto-public/pbcatalog/v2beta1/protocol_deepcopy.gen.go:2]: Another example of it producing empty-ish files.

**Comment by thisisnotashwin** [comment]
> For some of these comments that basically strings.Join all of the string names of the enum fields, is there a way to ensure that the comment and the enum don't drift with respect to each other?

That is a great question. I havent had the time to solve that. It is important for that comment to be up-to date from the kubebuilder perspective. I can work on it for 1.18 though, because i do think it is vital.

**Review by zalimeni — APPROVED** [review]
LGTM! Some minor comments but the generated code looks correct, and we'll know soon enough if it's not. Since this is only impacting new v2 resources, change looks safe to me to iterate on as needed.

**Pull request merged**